#!/bin/bash
#SBATCH -t 1000:00:00
#SBATCH --nodes=1
#SBATCH --export=NONE
#SBATCH	-o /data/marine_diseases_lab/erin/2020_Hemolymph_Dermo_Transcriptome_Project/SCRIPTS/SCRIPT_out_error_files/Prep_DESeq2_out_3_15_2021
#SBATCH	-e /data/marine_diseases_lab/erin/2020_Hemolymph_Dermo_Transcriptome_Project/SCRIPTS/SCRIPT_out_error_files/Prep_DESeq2_error_3_15_2021

# This script converts Stringtie output into format for use with DESeq2.
# Generates two CSV files containing the count matrices for genes and transcripts, using the coverage values found in the output of stringtie -e
# Python script (prepDE.py) provided by Pertea et al. to extract this read count information directly from the
# files generated by StringTie (run with the -e parameter)
# Example text file format:
	# ERR188021 <PATH_TO_ERR188021.gtf>

echo "Start $(date)"
module load python/2.7.6

# Create variable for each path to trimmed and quality filtered data folder
D=/data/marine_diseases_lab/erin/2020_Hemolymph_Dermo_Transcriptome_Project/ANALYSIS_FILES/C_vir_HISAT_STRINGTIE
PM=/data/marine_diseases_lab/erin/2020_Hemolymph_Dermo_Transcriptome_Project/ANALYSIS_FILES/Dermo_HISAT_STRINGTIE
S=/data/marine_diseases_lab/erin/2020_Hemolymph_Dermo_Transcriptome_Project/SCRIPTS

# Hemocyte transcriptomes
cd $D/
for i in *.merge.gtf; do
	# create text file with sample IDs and respective paths
	echo "$(echo $i |sed "s/\..*//") $D/$i" >> C_vir_Hemocyte_sample_list.txt
done

python $S/prepDE_Oct.2019.py -v -i C_vir_Hemocyte_sample_list.txt -g Hemocyte_2021_gene_count_matrix.csv -t Hemocyte_2021_transcript_count_matrix.csv
echo "C_vir_hemocyte DONE $(date)"

# Dermo mapped transcriptomes
cd $PM/
for i in *.merge.gtf; do
	# create text file with sample IDs and respective paths
	echo "$(echo $i |sed "s/\..*//") $PM/$i" >> C_vir_Dermo_2021_sample_list.txt
done

python $S/prepDE_Oct.2019.py -v -i C_vir_Dermo_2021_sample_list.txt -g Dermo_2021_gene_count_matrix.csv -t Dermo_2021_transcript_count_matrix.csv
echo "C_vir_Dermo DONE $(date)"

echo "All prepped $(date)"
